### PR TITLE
Add option to skip using _ignored_source field for synthetic source

### LINKED ISSUE #112940: Introduce a setting to skip using `ignoredSource`
### Description

`IgnoredSource` uses special logic to handle ignored values and encodes them at indexing time to later be able to return reconstructed source including ignored field values. We experienced some issues with it, affecting the decoding logic, and decided to introduce a mechanism that allows us to disable its usage. We need to do that both at index and query time. This setting is meant to be used as a "last-resourt" option to deal with possible issues affecting document indexing or queries.

The default behavior should be such that `ignoredSource` is used both at indexing time and query time. Not using the field at indexing or query time should be possible by changing the setting value.

Issues have been observer in clusters running version 8.15.0 and 8.15.1. As a result we need to make sure the bug fix is backported to 8.15.2.

### PR DESCRIPTION
**Pull request by lkts**
Closes #112940.

### CODE REVIEW

**Comment by lkts** [comment]
Diff is mostly tests.

File: server/src/test/java/org/elasticsearch/index/mapper/IgnoredSourceFieldMapperConfigurationTests.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+            b.endObject();
+        };
+
+        var doc = mapperService.documentMapper().parse(source(inputDocument));
+        // Field was written.
+        assertNotNull(doc.docs().get(0).getField(IgnoredSourceFieldMapper.NAME));
+
+        String syntheticSource = syntheticSource(mapperService, inputDocument);
+        // Values are not loaded.
+        assertEquals("{}", syntheticSource);
```
Comment (kkrik-es) [inline server/src/test/java/org/elasticsearch/index/mapper/IgnoredSourceFieldMapperConfigurationTests.java:58]: Flip to false, verify that the field gets populated.

File: server/src/test/java/org/elasticsearch/index/mapper/IgnoredSourceFieldMapperConfigurationTests.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+            b.endObject();
+        };
+
+        var doc = mapperService.documentMapper().parse(source(inputDocument));
+        // Field is not written.
+        assertNull(doc.docs().get(0).getField(IgnoredSourceFieldMapper.NAME));
+
+        String syntheticSource = syntheticSource(mapperService, inputDocument);
+        // Values are not loaded.
+        assertEquals("{}", syntheticSource);
```
Comment (kkrik-es) [inline server/src/test/java/org/elasticsearch/index/mapper/IgnoredSourceFieldMapperConfigurationTests.java:98]: Ditto.

File: test/framework/src/main/java/org/elasticsearch/index/mapper/MapperServiceTestCase.java [hunk 3]
```diff
@@ -827,30 +828,35 @@ protected final String syntheticSource(DocumentMapper mapper, CheckedConsumer<XC
...
             try (DirectoryReader roundTripReader = wrapInMockESDirectoryReader(DirectoryReader.open(roundTripDirectory))) {
-                String roundTripSyntheticSource = syntheticSource(mapper, roundTripReader, doc.docs().size() - 1);
+                String roundTripSyntheticSource = syntheticSource(mapperService, roundTripReader, doc.docs().size() - 1);
                 assertThat(roundTripSyntheticSource, equalTo(syntheticSource));
                 validateRoundTripReader(syntheticSource, reader, roundTripReader);
             }
         }
     }
 
-    protected static String syntheticSource(DocumentMapper mapper, IndexReader reader, int docId) throws IOException {
```
Comment (kkrik-es) [inline test/framework/src/main/java/org/elasticsearch/index/mapper/MapperServiceTestCase.java:847]: I'd keep this flavor and pass empty `indexSettings` for the cases we don't care. That should reduce the churn for testing files..

**Review by kkrik-es — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/index/mapper/IgnoredSourceFieldMapper.java [hunk 4]
```diff
@@ -49,12 +52,24 @@ public class IgnoredSourceFieldMapper extends MetadataFieldMapper {
...
     public static final String NAME = "_ignored_source";
 
-    public static final IgnoredSourceFieldMapper INSTANCE = new IgnoredSourceFieldMapper();
-
-    public static final TypeParser PARSER = new FixedTypeParser(context -> INSTANCE);
+    public static final TypeParser PARSER = new FixedTypeParser(context -> new IgnoredSourceFieldMapper(context.getIndexSettings()));
 
     static final NodeFeature TRACK_IGNORED_SOURCE = new NodeFeature("mapper.track_ignored_source");
 
+    public static final Setting<Boolean> SKIP_IGNORED_SOURCE_WRITE_SETTING = Setting.boolSetting(
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/index/mapper/IgnoredSourceFieldMapper.java:59]: maybe add java doc on top of these public settings to explain what they do exactly?

File: server/src/main/java/org/elasticsearch/index/mapper/SourceLoader.java [hunk 5]
```diff
@@ -112,13 +113,15 @@ class Synthetic implements SourceLoader {
...
          */
-        public Synthetic(Supplier<SyntheticFieldLoader> fieldLoaderSupplier, SourceFieldMetrics metrics) {
+        public Synthetic(Supplier<SyntheticFieldLoader> fieldLoaderSupplier, IndexSettings indexSettings, SourceFieldMetrics metrics) {
             this.syntheticFieldLoaderLeafSupplier = fieldLoaderSupplier;
             this.requiredStoredFields = syntheticFieldLoaderLeafSupplier.get()
                 .storedFieldLoaders()
                 .map(Map.Entry::getKey)
                 .collect(Collectors.toSet());
-            this.requiredStoredFields.add(IgnoredSourceFieldMapper.NAME);
+            if (indexSettings.getSkipIgnoredSourceRead() == false) {
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/index/mapper/SourceLoader.java:122]: My understanding was that we only need to ignore `_ignore_source` at read time, like is done here. This to avoid the potential errors when synthesizing the source. An escape hatch. Do we need ignore `_ignore_source` at write time?

Comment (kkrik-es) [reply server/src/main/java/org/elasticsearch/index/mapper/SourceLoader.java:122]: I think it's better to add both, in case we have a bug at some point (...) with indexing. These settings won't be advertised, but we can give them to support in case we have a really urgent SDH (ingestion blocked etc).

Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/index/mapper/SourceLoader.java:122]: Ok, I see. An another escape hatch if for some reason we have a bug at index time.

**Comment by lkts** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/index/mapper/IgnoredSourceFieldMapper.java (hunk 4 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/index/mapper/IgnoredSourceFieldMapper.java:65]: the java doc is now empty

**Review by martijnvg — APPROVED** [review]
LGTM  - this approach is cleaner

File: server/src/main/java/org/elasticsearch/index/mapper/IgnoredSourceFieldMapper.java (hunk 4 above)
Comment (lkts) [reply server/src/main/java/org/elasticsearch/index/mapper/IgnoredSourceFieldMapper.java:66]: Yes, fixed.

**Comment by lkts** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.15|[](https://github.com/elastic/elasticsearch/pull/113065)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

File: server/src/main/java/org/elasticsearch/index/mapper/IgnoredSourceFieldMapper.java [hunk 6]
```diff
@@ -151,6 +178,64 @@ static NameValue decode(Object field) {
...
+    public static Set<String> ensureLoaded(Set<String> fieldsToLoadForSyntheticSource, IndexSettings indexSettings) {
+        if (indexSettings.getSkipIgnoredSourceRead() == false) {
+            fieldsToLoadForSyntheticSource.add(NAME);
+        }
+
+        return fieldsToLoadForSyntheticSource;
+    }
+
+    @Override
+    protected SyntheticSourceSupport syntheticSourceSupport() {
```
Comment (kkrik-es) [inline server/src/main/java/org/elasticsearch/index/mapper/IgnoredSourceFieldMapper.java:194]: Why is this needed, in addition to the function above?

Comment (kkrik-es) [reply server/src/main/java/org/elasticsearch/index/mapper/IgnoredSourceFieldMapper.java:194]: oof alright
